File: crates/mediagram-core/src/state/merge.rs
//! Reconciling what several devices say about the same viewing.
//!
//! Matches `web/src/state/merge.ts`, pinned by shared fixtures to prevent
//! lost positions or resurrected Continue entries.
//!
//! **Last writer wins, per row** — the record for *one title* with the
//! highest `updated_at`, not per device and not per document.
//!
//! **`watched` is the tombstone for `progress`.** Finishing a title deletes
//! its position and writes a completion at the same moment, so a device
//! that has never heard of the completion still holds a position, and
//! merging naively would hand it back. A completion at least as new as a
//! position therefore beats it. Watchlist, Kids and collections need no such
//! trick — each row carries its own `removed` flag, reconciled by `keep`
//! below the same as any other.

use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use super::record::{CollectionRow, ListRow, ProgressRow, SyncRecord, WatchedRow, normal_name};

mod tie_break;
use tie_break::{Held, keep};

/// Everything the devices agree on, once they have been reconciled.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MergedProfile {
    /// The normalised name, which is what identifies a viewer across
    /// machines.
    pub name: String,
    /// The name as somebody actually typed it. Carried separately because
    /// the identity is normalised and a name is not.
    pub display_name: String,
    /// A kids profile if any device's document says so.
    #[serde(default, skip_serializing_if = "crate::state::record::is_false")]
    pub kids: bool,
    // `#[serde(default)]` throughout: a fixture's `expect` need only name
    // the fields it is testing, the same tolerance `record.rs` has.
    #[serde(default)]
    pub progress: Vec<ProgressRow>,
    #[serde(default)]
    pub watched: Vec<WatchedRow>,
    #[serde(default)]
    pub watchlist: Vec<ListRow>,
    #[serde(default)]
    pub collections: Vec<CollectionRow>,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct MergedState {
    pub profiles: Vec<MergedProfile>,
    /// Not scoped to a profile — see `schema.rs` on why `kids` alone has
    /// none.
    #[serde(default)]
    pub kids: Vec<ListRow>,
}

struct ViewerState {
    display_name: String,
    /// The device `display_name` was taken from.
    name_from: String,
    kids: bool,
    progress: HashMap<String, Held<ProgressRow>>,
    watched: HashMap<String, Held<WatchedRow>>,
    watchlist: HashMap<String, Held<ListRow>>,
    collections: HashMap<String, Held<CollectionRow>>,
}

/// Merges every device's document into one answer. Order-independent by
/// construction: merging A then B gives what merging B then A gives —
/// devices see each other's documents in whatever order Telegram hands them
/// over.
pub fn merge_states(records: &[SyncRecord]) -> MergedState {
    let mut by_viewer: HashMap<String, ViewerState> = HashMap::new();
    // Kids sits at the top level, not per viewer — see `schema.rs`.
    let mut kids: HashMap<String, Held<ListRow>> = HashMap::new();

    for record in records {
        let device = record.device.as_str();
        for row in &record.kids {
            keep(&mut kids, row.set_id.clone(), row.clone(), device);
        }

        for profile in &record.profiles {
            let Some(name) = normal_name(&profile.name) else {
                continue;
            };

            let held = by_viewer.entry(name).or_insert_with(|| ViewerState {
                display_name: profile.name.trim().to_string(),
                name_from: device.to_string(),
                kids: false,
                progress: HashMap::new(),
                watched: HashMap::new(),
                watchlist: HashMap::new(),
                collections: HashMap::new(),
            });
            // Device id determines spelling as well as row ties, independent
            // of the order in which documents arrive.
            if device > held.name_from.as_str() {
                held.display_name = profile.name.trim().to_string();
                held.name_from = device.to_string();
            }
            // Sticky: a document lacking the flag — an older device's —
            // cannot undo another device's word that this viewer is a kids
            // profile.
            if profile.kids {
                held.kids = true;
            }

            for row in &profile.progress {
                keep(&mut held.progress, row.set_id.clone(), row.clone(), device);
            }
            for row in &profile.watched {
                keep(&mut held.watched, row.set_id.clone(), row.clone(), device);
            }
            for row in &profile.watchlist {
                keep(&mut held.watchlist, row.set_id.clone(), row.clone(), device);
            }
            // A collection is one row on the wire, kept by its id: the later
            // edit wins outright, not item by item.
            for row in &profile.collections {
                keep(&mut held.collections, row.id.clone(), row.clone(), device);
            }
        }
    }

    let mut profiles = Vec::with_capacity(by_viewer.len());
    for (name, held) in by_viewer {
        let watched: Vec<WatchedRow> = held.watched.into_values().map(|h| h.row).collect();
        let finished_at: HashMap<&str, f64> = watched
            .iter()
            .map(|row| (row.set_id.as_str(), row.updated_at))
            .collect();

        let progress: Vec<ProgressRow> = held
            .progress
            .into_values()
            .map(|h| h.row)
            // The tombstone rule. `>=` rather than `>`: the two writes
            // happen in one moment and can carry the same millisecond, and
            // in a tie the completion is the later intention.
            .filter(|row| {
                finished_at
                    .get(row.set_id.as_str())
                    .copied()
                    .unwrap_or(-1.0)
                    < row.updated_at
            })
            .collect();

        profiles.push(MergedProfile {
            name,
            display_name: held.display_name,
            kids: held.kids,
            progress,
            watched,
            watchlist: held.watchlist.into_values().map(|h| h.row).collect(),
            collections: held.collections.into_values().map(|h| h.row).collect(),
        });
    }
    MergedState {
        profiles,
        kids: kids.into_values().map(|h| h.row).collect(),
    }
}

File: web/src/state/merge.ts
/**
 * Reconciling what several devices say about the same viewing.
 *
 * Every mistake here is silent. A merge that picks the older of two positions
 * loses an evening's watching and reports nothing; one that resurrects a
 * finished title puts it back on the Continue shelf for ever. So this is pure,
 * it is tested before anything is ever sent, and the rules are written down.
 *
 * **Last writer wins, per row.** Not per device and not per document: the
 * record for *one title* with the highest `updatedAt` is the one that counts.
 * Watching S1E4 on the phone and S1E9 on the laptop leaves both correct.
 *
 * **A completion is the tombstone for `progress`.** Finishing a title deletes
 * its position and writes a completion at the same moment — `clearProgress`
 * and `setWatched` are called together, and the v4 migration exists because
 * finishing would otherwise erase every trace. So a device that has never
 * heard of the completion still holds a position, and merging naively would
 * hand it back and put a finished film on the Continue shelf. A completion at
 * least as new as a position therefore beats it.
 *
 * **An `UnwatchedRow` removal is not a completion and does not, by itself,
 * suppress `progress`.** Un-marking says "not finished", not "finished at
 * this moment" — the player's own un-mark already leaves `progress` alone
 * (`store.ts`'s `setWatched`). A removal instead carries `lastFinishedAt`,
 * the completion it took the mark from, and a position is compared against
 * that: one no newer stays suppressed, a genuine rewatch made since
 * survives. See `watched-reconcile.ts` for how a live row and its removal
 * are weighed against each other, on a different key for a reason
 * `sync-record.ts` explains.
 */

import {
  normalName,
  type CollectionRow,
  type ListRow,
  type ProgressRow,
  type SyncRecord,
  type UnwatchedRow,
  type WatchedRow,
} from "./sync-record";
import { reconcileWatched } from "./watched-reconcile";
import { keep, type Held } from "./tie-break";

/** Everything the devices agree on, once they have been reconciled. */
export interface MergedProfile {
  /** The normalised name, which is what identifies a viewer across machines. */
  name: string;
  /**
   * The name as somebody actually typed it.
   *
   * Carried separately because the identity is normalised and a name is not:
   * a machine meeting a viewer for the first time creates a local profile,
   * and creating it from the identity would greet them as "andré". The first
   * spelling seen wins, which is arbitrary between "André" and "ANDRÉ" and
   * right in the only case that matters — there being just one.
   */
  displayName: string;
  /** A kids profile if any device's document says so; absent otherwise. */
  kids?: true;
  progress: ProgressRow[];
  watched: WatchedRow[];
  /** `mergeStates` always fills these; optional only so a hand-built
   * `MergedState` — a test, or `importMerged`'s caller — need not repeat an
   * empty array for every kind it has nothing to say about. */
  unwatched?: UnwatchedRow[];
  watchlist?: ListRow[];
  collections?: CollectionRow[];
}

export interface MergedState {
  profiles: MergedProfile[];
  /** Not scoped to a profile — see `schema.ts` on why `kids` alone has none. */
  kids?: ListRow[];
  /** Household-wide too, and kept per title the same way; see `schema.ts` v8. */
  editorsChoice?: ListRow[];
}

/**
 * Merges every device's document into one answer.
 *
 * Order-independent by construction: merging A then B gives what merging B
 * then A gives, which matters because devices see each other's documents in
 * whatever order Telegram hands them over.
 */
export function mergeStates(records: SyncRecord[]): MergedState {
  /** Per viewer, the best row seen so far for each title. */
  const byViewer = new Map<
    string,
    {
      displayName: string;
      /** The device `displayName` was taken from, for the tie-break below. */
      nameFrom: string;
      kids: boolean;
      progress: Map<string, Held<ProgressRow>>;
      watched: Map<string, Held<WatchedRow>>;
      unwatched: Map<string, Held<UnwatchedRow>>;
      watchlist: Map<string, Held<ListRow>>;
      collections: Map<string, Held<CollectionRow>>;
    }
  >();
  // Kids sits at the top level, not per viewer: marking a title as a
  // child's is a fact about the title, the same reason it has no profile in
  // `schema.ts`.
  const kids = new Map<string, Held<ListRow>>();
  const editorsChoice = new Map<string, Held<ListRow>>();

  for (const record of records) {
    const device = typeof record?.device === "string" ? record.device : "";
    for (const row of record?.kids ?? []) keep(kids, row.setId, row, device);
    for (const row of record?.editorsChoice ?? []) keep(editorsChoice, row.setId, row, device);

    for (const profile of record?.profiles ?? []) {
      const name = normalName(profile.name);
      if (name === null) continue;

      let held = byViewer.get(name);
      if (held === undefined) {
        held = {
          displayName: profile.name.trim(),
          nameFrom: device,
          kids: false,
          progress: new Map(),
          watched: new Map(),
          unwatched: new Map(),
          watchlist: new Map(),
          collections: new Map(),
        };
        byViewer.set(name, held);
      } else if (device > held.nameFrom) {
        // One viewer typed two ways on two devices. The spelling shown is
        // decided by device id, as a tie between rows is, so it does not
        // depend on which document Telegram happened to hand over first.
        held.displayName = profile.name.trim();
        held.nameFrom = device;
      }

      // Sticky: once any device calls this viewer a kids profile, no document
      // that merely lacks the flag — an older device's, say — can undo it.
      if (profile.kids === true) held.kids = true;

      for (const row of profile.progress ?? []) keep(held.progress, row.setId, row, device);
      for (const row of profile.watched ?? []) keep(held.watched, row.setId, row, device);
      for (const row of profile.unwatched ?? []) keep(held.unwatched, row.setId, row, device);
      for (const row of profile.watchlist ?? []) keep(held.watchlist, row.setId, row, device);
      // A collection is one row on the wire — the whole list, `removed`
      // included — so it is kept by its id rather than reconciled item by
      // item: two devices editing the same list within a merge round have
      // the later edit win outright, name and membership together.
      for (const row of profile.collections ?? []) keep(held.collections, row.id, row, device);
    }
  }

  const profiles: MergedProfile[] = [];
  for (const [name, held] of byViewer) {
    const { watched, unwatched, finishedAt } = reconcileWatched(
      new Map([...held.watched].map(([setId, one]) => [setId, one.row])),
      new Map([...held.unwatched].map(([setId, one]) => [setId, one.row])),
    );

    const progress = [...held.progress.values()]
      .map((one) => one.row)
      // The tombstone rule. `>=` rather than `>`: the two writes happen in one
      // moment and can carry the same millisecond, and in a tie the completion
      // is the later intention — a viewer who finished a title did not also
      // mean to leave a position in it.
      .filter((row) => (finishedAt.get(row.setId) ?? -1) < row.updatedAt);

    profiles.push({
      name,
      displayName: held.displayName,
      ...(held.kids ? { kids: true as const } : {}),
      progress,
      watched,
      unwatched,
      watchlist: [...held.watchlist.values()].map((one) => one.row),
      collections: [...held.collections.values()].map((one) => one.row),
    });
  }
  return {
    profiles,
    kids: [...kids.values()].map((one) => one.row),
    editorsChoice: [...editorsChoice.values()].map((one) => one.row),
  };
}
